jemalloc-sys fails to compile with clang 15

Reported by @mukilan with NixOS 23.05 in #31055:

```
error: failed to run custom build command for `jemalloc-sys v0.5.4+5.3.0-patched`
 
Caused by:
  process didn't exit successfully: `/home/mukilan/dev/servo/target/debug/build/jemalloc-sys-128f6f2e7894a2a7/build-script-build` (exit status: 101)
[...]
  src/malloc_io.c:107:9: error: incompatible pointer to integer conversion returning 'char *' from a function with result type 'int' [-Wint-conversion]
          return strerror_r(err, buf, buflen);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
  1 warning generated.
  1 warning and 1 error generated.
  In file included from make: *** [Makefile:478: src/malloc_io.sym.o] Error 1
  make: *** Waiting for unfinished jobs....
[...]
  thread 'main' panicked at /home/mukilan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/jemalloc-sys-0.5.4+5.3.0-patched/build.rs:351:9:
  command did not execute successfully: cd "/home/mukilan/dev/servo/target/debug/build/jemalloc-sys-0a3234e2aba5603a/out/build" && "make" "-j" "24"
  expected success, got: exit status: 2
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -190,6 +190,12 @@ xi-unicode = "0.3.0"
 xml5ever = "0.35"
 xpath = { path = "components/xpath" }
 
+# Force tikv-jemalloc-sys to build with at least -O1.
+# This can help with build environments that force _FORTIFY_SOURCE, such as Nix.
+# More details: <https://github.com/tikv/jemallocator/issues/108>
+[profile.dev.package.tikv-jemalloc-sys]
+opt-level = 1
+
 [profile.release]
 opt-level = 3
 debug-assertions = true

--- a/shell.nix
+++ b/shell.nix
@@ -5,7 +5,10 @@
   buildAndroid ? false
 }:
 with import (builtins.fetchTarball {
-  url = "https://github.com/NixOS/nixpkgs/archive/1e5b653dff12029333a6546c11e108ede13052eb.tar.gz";
+  # NixOS users: if servoshell crashes with an assertion failure in surfman’s x11/connection.rs,
+  # eglInitialize() may be failing, or you may be building with an incompatible version of glibc.
+  # Use your system nixpkgs here, change `llvmPackages` below if necessary, then do a clean build.
+  url = "https://github.com/NixOS/nixpkgs/archive/d88bee41dbec574b3c2bf97de37371d90f96475d.tar.gz";
 }) {
   overlays = [
     (import (builtins.fetchTarball {
@@ -28,11 +31,7 @@ let
       url = "https://github.com/NixOS/nixpkgs/archive/6adf48f53d819a7b6e15672817fa1e78e5f4e84f.tar.gz";
     }) {};
 
-    # We need clangStdenv with:
-    # - clang < 16 (#30587)
-    # - clang < 15 (#31059)
-    # - glibc 2.38 (#31054)
-    llvmPackages = llvmPackages_14;
+    llvmPackages = llvmPackages_20;
     stdenv = llvmPackages.stdenv;
 
     buildToolsVersion = "34.0.0";
@@ -63,6 +62,9 @@ in
 stdenv.mkDerivation (androidEnvironment // {
   name = "servo-env";
 
+  # NOTE: tshark(1) for etc/devtools-parser.py can’t be installed here, because it requires
+  # CAP_NET_RAW and CAP_NET_ADMIN. Install Wireshark with your system package manager,
+  # or for NixOS, enable the option: `programs.wireshark.enable = true;`
   buildInputs = [
     # Native dependencies
     fontconfig freetype libunwind
@@ -81,7 +83,6 @@ stdenv.mkDerivation (androidEnvironment // {
     llvmPackages.bintools # provides lld
 
     udev # Needed by libudev-sys for GamePad API.
-    wireshark-cli  # for `tshark` in etc/devtools_parser.py
 
     # Build utilities
     cmake dbus gcc git pkg-config which llvm perl yasm m4

